### ISSUE TITLE
[backend][연구] YOLO 모델 훈련 계획 수립

### ISSUE THREAD

**Issue by techgyu** [HDRT-team-2/HDRT-Final-Project#157]
# YOLO 모델 훈련 계획 수립

---

## 1. 연구 목적

시뮬레이터 기반 전장 상황 데이터셋 구축 완료에 따라,
향후 객체 탐지 모델 훈련을 위한 YOLO 학습 구조 및 절차 이해를 목적으로 합니다.
본 단계에서는 훈련 방법론, 데이터 구조, 하이퍼파라미터 설정 방식 등을 학습 및 정리합니다.

## 2. 학습 및 계획 수립 예시(아래를 참고하여 수립 계획에 대해 댓글로 기재바랍니다.)

---

### 1. YOLO 구조 이해

**입력 이미지 처리 흐름:**
YOLO는 입력 이미지를 고정된 크기(예: 640×640)로 리사이즈하고, 정규화 과정을 거쳐 모델에 입력합니다.
이후 네트워크는 이미지를 S×S 그리드로 분할하고, 각 셀에서 객체 존재 확률과 경계 상자(Bounding Box) 좌표를 동시에 예측합니다.

**Backbone / Neck / Head 역할:**
- Backbone: 이미지의 기본 특징(Edges, Shapes, Patterns)을 추출하는 부분입니다. (예: CSPDarknet, EfficientNet 등)
- Neck: 다중 스케일(feature pyramid)을 형성해 작은 객체와 큰 객체 모두 탐지할 수 있도록 특징을 통합합니다. (예: PANet, FPN 등)
- Head: 최종적으로 각 스케일에서 객체의 클래스 확률과 Bounding Box를 출력합니다. (YOLOv8 이후는 Anchor-free 구조 사용)

**Anchor 기반 탐지 방식 및 NMS 처리 과정:**
- Anchor Boxes: 사전에 정의된 다양한 크기·비율의 박스를 기준으로 객체를 탐지합니다.
- NMS(Non-Maximum Suppression): 여러 중복 박스 중 Confidence Score가 가장 높은 박스만 남기고 나머지를 제거하여 중복 탐지를 방지합니다.

### 2. 훈련 절차 학습

**데이터셋 디렉터리 구성:**

dataset/
├── images/
│   ├── train/
│   ├── val/
└── labels/
    ├── train/
    ├── val/

각 이미지와 동일한 이름의 .txt 파일을 생성하며, 한 줄에 객체 1개씩
[class_id, x_center, y_center, width, height] 형식으로 기록됩니다. (좌표는 0~1 정규화)

**data.yaml 및 cfg 설정 방법:**
- data.yaml: 데이터 경로, 클래스 수, 클래스 이름을 지정하는 파일입니다.
```yaml
train: ./dataset/images/train
val: ./dataset/images/val
nc: 8
names: ['person', 'vehicle', 'mine', ...]
```
- cfg: 모델 구조 및 하이퍼파라미터 설정 파일로, backbone 깊이·채널 수, optimizer, batch size 등을 포함합니다.

### 3. 주요 학습 인자 역할 이해:

- Epoch: 전체 데이터셋을 한 번 학습하는 주기 (예: 100 epochs → 데이터셋 100회 반복 학습)
- Batch size: 한 번에 학습되는 이미지 수 (메모리와 학습 안정성에 영향)
- Learning rate (lr): 가중치 업데이트 크기, 모델 수렴 속도 결정
- Optimizer: AdamW, SGD 등 가중치 조정 방식 선택
- Image size: 탐지 정확도와 속도에 영향을 주는 핵심 요소

### 4. 훈련 유형 비교

**Pretrained Weight 기반 Fine-tuning:**
- 장점: 빠른 수렴, 적은 데이터로도 높은 성능 가능
- 단점: 기존 데이터셋(BDD100K, COCO 등)에 맞춰진 bias가 남을 수 있음
- 적합 상황: 도메인이 유사하거나 데이터셋 크기가 작을 때

**신규 학습 (From Scratch):**
- 장점: 완전히 새로운 환경(예: 시뮬레이터 전장 데이터)에 최적화된 모델 생성 가능
- 단점: 훈련 시간 증가, 더 많은 데이터 필요
- 적합 상황: 도메인이 기존 데이터셋과 크게 다를 때 (예: 군사, 야간, 열상 등)

### 5. 성능 평가 기준 학습

- **Precision (정밀도):**
모델이 ‘탐지했다’고 한 것 중 실제로 맞은 비율
→ 거짓 양성(False Positive)이 적을수록 높음

- **Recall (재현율):**
실제 존재하는 객체 중 모델이 탐지한 비율
→ 거짓 음성(False Negative)이 적을수록 높음

- **mAP (mean Average Precision):**
Precision–Recall 곡선에서의 평균 면적값
mAP@50: IoU가 0.5 이상일 때의 평균 정확도
mAP@50-95: IoU 0.5부터 0.95까지 구간별 평균 (보다 엄격한 평가 기준)

- **Confusion** Matrix 분석:**
각 클래스별로 TP(정탐), FP(오탐), FN(미탐)을 시각적으로 표시하여
특정 클래스에서의 탐지 오류 경향을 파악할 수 있음.

**Comment by choongone** [comment]
labeling된 데이터들을 가지고 학습을 시작하는 과정까지 정리 한 내용은 아래와 같습니다. 

## 1. 데이터셋 구조 정리


project_root/
├── dataset/
│             ├── images/
│             │        ├── train/
│             │        └── val/
│             └── labels/
│                       ├── train/
│                       └── val/
├── data.yaml
└── yolo11x.yaml

이 형식으로 폴더 생성해놓기 


## 2. data.yaml 작성

train: ./dataset/images/train
val: ./dataset/images/val
nc: 8
names:
  [
   'Car 2',
   'Car 3', 
   'Car 4',
   'Human 1', 
   'Tank 1', 
   'Rock 1', 
   'Rock 2', 
   'Mine 1',
   'Wall 2', 
   'Wall 2  X 10',
   'Other'
  ]	


## 3. yolo11x.yaml 파일에 메모장으로 

backbone:
  - [Conv, 64, 3, 2]
  - [Conv, 128, 3, 2]
  - [C2f, 128, 3]
  - [Conv, 256, 3, 2]
  - [C2f, 256, 6]
  - [Conv, 512, 3, 2]
  - [C2f, 512, 6]
  - [SPPF, 512, 5]

neck:
  - [Conv, 256, 1, 1]
  - [Upsample, None, 2, nearest]
  - [Concat, [256, 256]]
  - [C2f, 256, 3]
  - [Conv, 128, 1, 1]
  - [Upsample, None, 2, nearest]
  - [Concat, [128, 128]]
  - [C2f, 128, 3]

head:
  - [Detect, [128, 256, 512]]  # Detection Head

클래스 수 (data.yaml의 nc 값과 동일해야 함)
nc: 8

하이퍼파라미터 설정
depth_multiple: 1.0
width_multiple: 1.0

작성


## 4. 학습 명령어에
project_root 폴더로 이동한 후 vscode 터미널로

yolo detect train model=yolo11x.yaml data=data.yaml epochs=300 imgsz=960 batch=16 weights=None

입력해서 학습 시작
